### DISCUSSION TITLE
Recording segment load times in Safari [iOS]

### DISCUSSION THREAD

**Discussion by ofcedwards** [blakeblackshear/frigate#13862] [Ask A Question] [answered]
Are there any known issues with recording playback in Safari on iOS? I have Frigate saved as a PWA and have noticed that remote connection to Frigate via VPN is so slow with recording playback that it's unusable. Opening Frigate in the Firefox iOS app alleviates this (and playback has no issues in any browser while on local wifi). As a side note, my video segments are typically around 10mb (8196 bitrate) and my home upload speed varies from 20mbit to 40mbit.

I've saved some screenshots of the web inspector network tab for each scenario. In both scenarios, I was connected via VPN and attempting to playback the same review item.

**Playback with PWA**
<img width="999" alt="Screenshot 2024-09-20 at 17 16 02" src="https://github.com/user-attachments/assets/bd9cbd88-49b7-4edb-91f5-bc2fe893b8e5">

**Playback with Firefox App**
<img width="996" alt="Screenshot 2024-09-20 at 17 16 55" src="https://github.com/user-attachments/assets/8c99f4eb-c5e7-4fb7-b578-cd2cdeaa91c3">

**Comment by NickM-27** [comment]
I have no issues using iOS PWA although I have a symmetrical gb fiber connection so upload is no concern. This is using a cloudflare tunnel directly to frigate

**Reply by ofcedwards** [reply]
It just seems strange to me that performance would vary that much between PWA and Firefox. My first inclination was a network speed issue, but I have a very short initial buffer with Firefox and then it runs continuously. The PWA (and Safari itself) have a massive initial buffer time and may only make it through one segment before stopping to buffer again.

It may also be worth noting that the network page shows times of ~6 seconds on Safari while connected via wifi (~800mbit connection between router and iPhone). Also, when on VPN and using Firefox, my Unraid server shows steady outbound traffic matching my upload speed. When using Safari, Unraid shows small bursts of outbound traffic, but does not seem to have a continuous flow.

**Reply by NickM-27** [reply]
there are a couple different possibilities. One is that safari supports hls directly so it is likely using the m3u8 native player meanwhile firefox doesn't so it is likely using hls.js, perhaps the safari browser is just struggling for some reason

**Reply by ofcedwards** [reply]
> This is using a cloudflare tunnel directly to frigate

Safari performance is significantly better when accessing Frigate through a cloudflare tunnel. Seems like this is going to be some issue with how wireguard is configured

**Reply by ofcedwards** [reply]
@NickM-27 I'm reopening this issue after doing some more digging. I went through many iterations of Wireguard configurations and could not do anything that would make a meaningful difference. Also, streaming through Wireguard to Safari on iPadOS and macOS (all latest release versions) works perfectly. It appears that Safari on iOS would regularly get stuck while attempting to load a video segment. It would eventually timeout/retry and after enough segments were loaded, playback would start until it hung on another segment. During these hangs, zero data was being downloaded to the client.

My initial tests have shown that this issue can be resolved by adding `keepalive_disable safari;` to nginx.conf in Frigate. After an initial buffer, playback starts and continues until the end of each hour where there's another short buffer. I can see that data is constantly being downloaded by the client as well. I have not noticed any performance hit from this change, but my Frigate instance is fairly small and does not have multiple concurrent users. Not having a good grasp of the bigger picture, I'll leave it up to you to decide if this change should be added into Frigate.

**Reply by NickM-27** [reply]
At what level did you apply this? Presumably if it's applied only at the `vod` level it should fix the issue without affecting other areas

**Reply by ofcedwards** [reply]
I added it to the `http` block directly under `keepalive_timeout 65;`

Looks like this can be done by location. I'll have to spend some more time on it narrow it down only to where it's needed

**Reply by NickM-27** [reply]
I'd assume it's only needed in the vod location

**Reply by ofcedwards** [reply, marked as answer]
Fixed by #13909
